PFW-1376 MMU error during mid-print does not pause/resume print

gudnimg opened the pull request:
JIRA ticket: https://dev.prusa3d.com/browse/PFW-1376

* If the MK3S reports an error in the middle of a print move then we want to pause the print until the error is resolved. The pausing and parking needs to be done in two main `loop()` iterations which is why I have split all the actions into switch-case.
*  If the MMU recovers from a communication timeout _without_ the user interacting with the LCD screen, the error screen will be dismissed and the print resumed automatically.

wavexx committed:
Unify fullscreen/multiscreen/yes-no prompt message handling

Modify lcd_show_multiscreen_message_two_choices_and_wait_P to also
handle single-screen or empty (no-clear) prompts, making other functions
redundant. Saves 76 bytes.

Change existing functions to simply call
lcd_show_multiscreen_message_two_choices_and_wait_P with the correct
arguments.

This changes the prompt of existing Yes/No messages: the previous prompt
would use the last two lines of the LCD, while the new prompt is using
just the last line of the LCD instead.

Translation do not require updates, since the Yes/No translation was
already the same in both implementations.

wavexx committed:
Allow choice positioning. Balance Yes/No messages.

Add an additional parameter to control the position of second choice
prompt position (while defaulting to the old).

This allows Yes/No prompts to be equally spaced.

leptun committed:
Use PAT9125 new init sequence

leptun committed:
Unflip the hardcoded flip of the Y axis

leptun committed:
Filament sensor refactoring initial

leptun committed:
Fix repeated autoload menu

leptun committed:
IR volt debugging

leptun committed:
Remove old fsensor init

leptun committed:
Remove OQ and add fancy autoload interaction

leptun committed:
IR_sensor_analog voltage readings

leptun committed:
Remove old fsensor_init()

leptun committed:
Remove dead code

leptun committed:
More dead code removal

leptun committed:
Move more IR sensor analog stuff to Filament_sensor.h

leptun committed:
Remove unused fsensor variables

leptun committed:
Checkpoint after a lot of removal

leptun committed:
Revert menu hack

leptun committed:
Fsensor error state

leptun committed:
Separate runout and enabled settings

leptun committed:
Enable fsensor when preparing for shipping

leptun committed:
Remove original marlin fsensor code

leptun committed:
Correct ifdefs for all IR sensor types

leptun committed:
Fix MK2 build?

leptun committed:
fsensor add deinit

leptun committed:
fsensor event blanking

leptun committed:
Fix reinit on error

leptun committed:
Add base code for PAT9125_sensor

leptun committed:
Cleanup pat9125.h

leptun committed:
swi2c safe init

leptun committed:
Use _check instead of register read for probe()

leptun committed:
twi safe init

leptun committed:
Add disable() to swi2c as well

leptun committed:
Shorter TWI timeout

leptun committed:
Correct sensor orientation for old pat9125 init

leptun committed:
Update Sensor info menu documentation

leptun committed:
PAT9125 first prototype

leptun committed:
Checkpoint: jam detection

leptun committed:
Improve PAT9125 UI implementation

leptun committed:
Fix build (??)

leptun committed:
Merge branch 'MK3' into MK3_MMU2

Commit:
MMU2 interface overhaul

First port of the new MMU2-printer interface into 8bit FW.

Commit:
Prune most of TODO from mmu2.cpp

+ refactor sound.h for better readability

Commit:
Return ir_sensor_detected + related functionality

Commit:
Implement mmu2 serial interface

Commit:
Avoid exposing internal variables of fsensor

makes the code a bit shorter as well, because the compiler can be sure
that the variables are not accessed from the outside of the
fsensor.cpp compilation unit.

leptun committed:
Merge pull request #3423 from leptun/MK3_PAT9125_NEW_INIT

PAT9125 new init sequence

Commit:
Refactor Tx/Tc/T? commands

leptun committed:
Merge pull request #3429 from leptun/MK3_Filament_sensor_refactoring

Filament sensor refactoring

Commit:
Fix build after rebase

Commit:
Report state of fsensor into the MMU code

leptun committed:
Implement fsensor raii event suppression

Commit:
Introduce MMU errors + their texts

Commit:
Add original M600 MMU handling

seems to be almost unrelated to MMU function at all,
so returning it almost intact.

Commit:
Add MMU FW version variables

hardcoded for now, but API ready

Commit:
Add ProgressCodes -> text converter

leptun committed:
Remove old mmu.cpp

leptun committed:
Implement MMU hw reset

gudnimg committed:
Fix crash in manage_response

The watchdog needs to be kicked in the for(;;) loop.
This is usually done by calling manage_heater before
manage_inactivity.

gudnimg committed:
Fix PROGMEM read in ProgressCode message

This could cause the printer to crash. The Serial Stream would show
a garbled string and the same corruption would appear on the Status
Screen's status line.

gudnimg committed:
Fix overwrite in ProgressCode message buffer

When snprintf() if called we need to read the return value to see how
many bytes were written. Then when we call strncpy_P() through
TranslateProgress(), we need to tell the code to start writing
at byte 'len', or &msg[len]. Also we need to update the byte size
which strncpy_P() is allowed to write (64 - len).

leptun committed:
Increase RX buffer size

gudnimg committed:
Fix PROGMEM read in ErrorCode message

This could cause the printer to crash. The Serial Stream would show
a garbled string.

Also included in this commit:
* The buffer size is increased from 64 bytes to 192 bytes.  We need to
take into account the length of the ErrorTitle and ErrorDescription.
* Fix overwrite in ErrorCode message buffer.

gudnimg committed:
Add missing space in ProgressCode message

Such that:
MMU2:P=123EngageIdler
Becomes:
MMU2:P=123 EngageIdler

gudnimg committed:
Add missing spaces to ErrorCode messages

Such that:
MMU2:E=32766ErrorTitleTextDescription
Becomes:
MMU2:E=32766 ErrorTitle TextDescription

Also simplified the process of combining ErrorTitle and TextDescription
into the msg buffer by using snprintf. This is saver since we only
use dstSize in one place instead of two.

gudnimg committed:
Add a read function for mmu_print_saved
and remove the old extern value which is not used anymore

I had to remove "MMU2::" from mmu2_protocol_logic.cpp, otherwise the build failed.

gudnimg committed:
Fix return value of WhereIsFilament()

getFilamentPresent returns the state of the IO pin of the sensor.
So we should use AT_FSENSOR instead of IN_NOZZLE
since the fsensor does not know if the filament is actually in the nozzle.

gudnimg committed:
Move extruder when loading filament to fsensor

When the fsensor see filament, stop the extruder

gudnimg committed:
Merge pull request #1 from wavexx/unify_fullscreen_msg_handling

Merge PR-3090

gudnimg committed:
add that downwards facing >> character

gudnimg committed:
Print the MMU progress into status screen

gudnimg committed:
Remove class LcdUpdateDisabler

Calling lcd_update_enable directly gives us more control

This also save a bit of flash memory

gudnimg committed:
Allow LCD to be updated in manage_response loop

Also removed one TODO which David mentioned is not needed anymore

gudnimg committed:
Update lcd_show_multiscreen_message_two_choices_and_wait_P

to support 3 choices

gudnimg committed:
Don't disable LCD updates in menu_item_function_P

gudnimg committed:
Implement PFW-1312 in MMU code

gudnimg committed:
Checkpoint:

- Rename lcd_show_multiscreen_message_two_choices_and_wait_P to lcd_show_multiscreen_message_with_choices_and_wait_P

- Move MMU error screen from ReportError() to ReportErrorHook()

- Fix the menu selection to work for menus with either two choices or three choices

- The buttons shown on the MMU error menu are now determined by the definition of btnOperation[]

gudnimg committed:
Remove unused includes after I moved the MMU error screen

gudnimg committed:
Add helper macros to parse nibbles

Also made the button operations and nibbles constant since they should not be modified.

gudnimg committed:
Checkpoint

- combined lcd_show_three_choices_prompt_P and lcd_show_two_choices_prompt_P into one function called lcd_show_choices_prompt_P
- Removed third_col parameter
- Added an enum to make the button selection result more readable

Commit:
Update error codes + constexpr FindErrorIndex

Beware - requires avr-gcc 7.3!

gudnimg committed:
Make all choice menus return consistent values

Such that left = 0, middle = 1, and right = 2

gudnimg committed:
Fixup

gudnimg committed:
Merge branch 'mk3-mmu' into PFW-1312

gudnimg committed:
Update the error menu after merging new changes

Will test this tomorrow

gudnimg committed:
Change name of variable 'yes' to be more descriptive

gudnimg committed:
Change default selection parameter from bool to uint8_t

This allows us make any of the three button choices the default selection

gudnimg committed:
Initial commit for loading filament to the nozzle

Commit:
Fix compile-time conversion MMU2 ErrorCode -> Prusa-Error-Code

and hide some of the implementation details into mmu2_error_converter.cpp
which makes the code in mmu_reporting.cpp much easier to read.

Commit:
Make sure the MMU error title never overflows on the LCD

Commit:
Fix FW update needed button order

DRracer committed:
Merge pull request #15 from gudnimg/PFW-1312

PFW-1312 Define layout of LCD for MMU error screens

3d-gussner committed:
C++17 Arduino_boards version 1.0.5-1

gudnimg committed:
Improve load to nozzle sequence

- Fixed a bug where the feed rate was 60x higher than intended
- Adjusted the sequence via trial and error. The sequence where the filament has not exited the nozzle is 35mm shorter than the previous implementation
- I reduced the feed rate slightly (by 1mm/s) on the fast load sequence

gudnimg committed:
Reduce grinding when loading to the fsensor

- We need to take into account additional 35mm move by the MMU.
- While the fsensor is not triggered, reduce the extruder movement each step from 20mm to 5mm. This should help reduce the variance in the filaments position before we run the final load to nozzle sequence.

gudnimg committed:
Add the load to nozzle sequence once the MK3S takes control and MMU has disengaged idler.

Also removed ProgressCode::FeedingToNozzle from switch-case since it is not used for this purpose at the moment.

gudnimg committed:
Add TODO to prompt user if extruder is extruding correct color

gudnimg committed:
Don't block OnMMUProgressMsg while waiting for fsensor to trigger

gudnimg committed:
Add back menu it confirm if extrusion has correct color after loading to nozzle via LCD

I've added the call to this message via the menu function to ensure it
only appears when interacting with the LCD

gudnimg committed:
Fix default selection after changes in PFW-1312

The Yes and No are reverted. Before PFW-132 Yes is 1 and No is 0
Now they are Yes is 0, and No is 1

gudnimg committed:
Remove redundant call to lcd_update_enable

gudnimg committed:
Remove unused parameter nlines to simplify things

This allows us to remove one overload
of lcd_display_message_fullscreen_P

gudnimg committed:
fixup

Commit:
Update MK3S to match MMU PR#168

- new progress and error codes
- remove #error from mmu2_power.cpp as it makes no sense
- remove temporary error translation code

Commit:
Reorder errors list according to Prusa-Error-Codes #61

Commit:
Handle the new MMU states (slow load to fsensor)

Commit:
This starts to print ;)

Temporarily disabled filament runout as this breaks ToolChange operation
yet for unknown reason.

Commit:
Fix compilation MK3

Commit:
Fix build MK25S

Commit:
Fix MK25 build

gudnimg committed:
PFW-1318 Implement callback to render status line

To render the thermometer and degree symbols
the CGRAM table will instead swap out
the feedrate and clock symbol when
lcd_set_custom_characters_nextpage()
is called.

Commit:
Suppress communication drop-out < 8s (MMU in bootloader)

PFW-1327

Commit:
Convert button on LCD into button for MMU

- Dependent on current MMU error being processed
- Includes rename of LEFT_BUTTON_CHOICE -> LCD_LEFT_BUTTON_CHOICE (and right, middle as well) to avoid confusion with MMU buttons' ordering

Commit:
Update printer's type (MMU/non-MMU) at runtime

+ reformat utils.cpp

Commit:
Fix lang build

Commit:
Fix build

Commit:
Reuse FSensor runout blocking impl. throughout the whole FW

gudnimg committed:
Merge branch 'mk3-mmu' into PFW-1318

Fix merge conflict

gudnimg committed:
Fix compile error

gudnimg committed:
PFW-1318 Rewrite the implementation to be non-blocking

gudnimg committed:
Cleanup: Remove unused callback

Remove unused callback in lcd_show_multiscreen_message_with_choices_and_wait_P

gudnimg committed:
Cleanup: Add comments and reduce code size slightly

gudnimg committed:
Call lcd_set_custom_characters_nextpage

gudnimg committed:
Fix issue where error screen updates too slowly in manage_response

gudnimg committed:
Update printer name if M862.3 is called but not M862.2

gudnimg committed:
PFW-1323 Fix MMU error screen doesn't disappear when error resolved on the MMU

gudnimg committed:
Cleanup: Pull out PrusaErrorCodeIndex

Saves 8 bytes of flash

gudnimg committed:
Cleanup: fix parameter comments

gudnimg committed:
Cleanup: Fix two compiler warnings

gudnimg committed:
Cleanup: Remove an old comment

DRracer committed:
Merge pull request #19 from gudnimg/PFW-1318

PFW-1318 and PFW-1323

Commit:
Hack around old MMU FW's msgs

Commit:
Re-enable filament runout + add debug msgs

Needs a fix on the MMU side (report FeedToBondtech::DisengageIdler)

Commit:
Reset MMU upon start + fix reporting its start errors

Commit:
Make init MMU error screens send buttons (Idle mode)

gudnimg committed:
PFW-1336

The following situation is now fixed, steps:
1. Load filament through LCD
2. Load to nozzle through LCD
3. Before filament reaches Fsensor => turn off the printer
4. Wait a few seconds and then turn the printer on again
5. Start a print through the SD card
6. Printer will unload and then re-load the filament

Previously in step 6, the printer would start a print without any
filament loaded.

vintagepc committed:
- Fix DisableMMU button on FW incompat
5823c14
- Implement eeprom var, fix reset

vintagepc committed:
Fix wrong return

vintagepc committed:
Allow reset if the EEPROM is set to "on"

vintagepc committed:
Add missing CheckUserInput calls, fix bad array entry for FW_update

DRracer committed:
Merge pull request #20 from gudnimg/fix_sPrinterName

Fix sPrinterName not set when using MMU

Commit:
Fix build MK25S

vintagepc committed:
Fix some partking and recovery bugs

Commit:
Temporarily disable runout

There are still printers which suffer from spurious runouts for no reason.
For testing purposes, runout will be disabled until the real cause is found.

gudnimg committed:
Disable filament runout when MMU is unloading

Commit:
Disable filRunout when MMU.Enabled()

It turned out the runout is caused by the ramming sequence in the G-code, which is interpreted before the actual MMU command gets into processing.
The frequency of these errors/runouts differed one machine from another and was really hard to track down.
Therefore - once the MMU is enabled (active and running) we'll disable filament runout handling caused by the printer's filament sensor.

gudnimg committed:
Disable automatic in M600

gudnimg committed:
Introduce a way to read the previous tool used

gudnimg committed:
During Tx, temporarily allow cold extrusion

This prevents a FSENSOR DIDNT TRIGGER error
and allows the bondtech gears to grab the filament.

gudnimg committed:
Add Unloading message to LCD

Also made mmu_M600_load_filament static to limit its scope

gudnimg committed:
Minor cleanup

gudnimg committed:
Implement 'Filament not loaded' option for M600 with MMU

gudnimg committed:
PFW-1343 initial commit

vintagepc committed:
First pass, improving the error recovery.

vintagepc committed:
Button handling WIP

vintagepc committed:
Fixed buged temperature resume.

gudnimg committed:
Write documentation for lcdui_print_extruder

vintagepc committed:
Fix button enum

vintagepc committed:
fix bug I introduced

vintagepc committed:
remove logspam

gudnimg committed:
PFW-1344 Fullscreen message Loading Filament

When loading a filament through the LCD,
show a full screen message:

Loading Filament N

Where N can range from 0 to 5.

3d-gussner committed:
Update to Arduino_boards version 1.0.5-1

Commit:
Add a Try-Load-to-Bondtech operation

PFW-1347

Commit:
Report detected MMU version in the Support menu

It was just subtly broken, all the functionality had already been implemented.

Commit:
Bump required MMU FW version to 2.0.18

Commit:
Avoid drawing MMU Error screen while MMU moves manually

Commit:
Fix rendering temp. + toolchange on MMU error screen

Commit:
Disable fsensor newVolt debug reporting

Commit:
Perform 3xRetry automatically

This is a draft PR showing the potential 3x retry implementation on the printer's side.
It is much less code and looks more reliable than the same functionality in the MMU FW.

Still, more work needs to be done:
- [ ] Button is sent to the MMU even before returning from the parking position
- [ ] Then the button is sent again
- [ ] Then the printer runs out of retryAttempts

We need to find a better spot to check for "automatic" retry and issuing of the buttons

gudnimg committed:
Partial fix for PFW-1334

When a FSENSOR error happens, the first step for the MK3S should be to stop any loading by stopping the E-motor.

From this point, the buttons should determine what the MK3S does next.

gudnimg committed:
PFW-1334 Move E-motor on FSENSOR error retry

gudnimg committed:
Add MMU2_UNLOAD_TO_FINDA_FEED_RATE

gudnimg committed:
PFW-1334 Distinguish between a retry unload and U0 unload

U0 unload uses the ramming sequence, while the retry unload does not because the MMU starts unloading much sooner

vintagepc committed:
PFW-1351 Cooldown timeout (#33)

* PFW-1351 WIP

* Fix bug with cooldown timing due to bad bitfield.

* Use default safety timer value for timeout

gudnimg committed:
Fix PFW-1357 (#35)

lcd_encoder_diff would be set to 0 if lcd_update_enable(true)
is called outside ReportErrorHookMonitor
This would put enc_dif out of sync with lcd_encoder_diff and could cause false rotation triggering

gudnimg committed:
Merge branch 'mk3-mmu' of https://github.com/DRracer/Prusa-Firmware into mk3-mmu

vintagepc committed:
An attempt at fixing the retry. Discussion needed.

gudnimg committed:
PFW-1340 Hide Done button for two errors

* FSENSOR_DIDNT_TRIGGER
* FSENSOR_DIDNT_GO_OFF

The Done button does not Move the E-motor because it
expects the user to have manually resolved the problem

Also if the filament is in the gears, we cannot pull the filament out.
In this case the Retry button is more suited as it will unload the filament.

3d-gussner committed:
Merge pull request #39 from gudnimg/remove-done-button-two-errors

PFW-1340 Hide Done button for two errors

gudnimg committed:
* Correct the unload length on the MK3S side
* Pull all numbers into global constants

vintagepc committed:
Fix trying to use serial before it's intiialized

vintagepc committed:
Merge pull request #41 from DRracer/fix-mmu-startup

Fix trying to use serial before it's intiialized

gudnimg committed:
Optimise mmu_cut_filament_menu (#40)

preheat_or_continue should be used here to reduce copying code

gudnimg committed:
Add a missing break

I don't see any difference in behaviour but this should be more correct

gudnimg committed:
PFW-1373

Fix the unload procedure when the user has paused a print
then stopped the print after the temperature has reached below 175°C
Now the E-motor will move as expected

3d-gussner committed:
Merge pull request #36 from gudnimg/stop-e-motor-on-error

PFW-1334 Move E-motor on FSENSOR error retry

3d-gussner committed:
Update build.sh as in MK3 branch

3d-gussner committed:
Merge pull request #44 from 3d-gussner/mk3-mmu_build

Update build.sh as in MK3 branch

gudnimg committed:
PFW-1375 MMU error reported by MK3S does not appear on LCD

gudnimg committed:
Merge branch 'mk3-mmu' into PFW-1375

gudnimg committed:
PFW-1354 Fix slow button processing (#37)

* Fix MK3S buttons being processed too slowly

After a button is selected on the MK3S MMU error screen, we need to process it before asking for another MMU Query.

Because the MMU Query will make the Error screen return and overwrite the status screen

* Add back CheckUserInput for ButtonPushed event

gudnimg committed:
Improve Communication Timeout screen

Automatically dismiss the Communication timeout error screen
if the MMU becomes responsive again.

Example of this is if the MMU power connector was unplugged,
and then plugged back again.

gudnimg committed:
Initial commit for PFW-1376

If an error happens during a print, save the print
and recover the print once the error is gone.

gudnimg committed:
Fix merge conflict

gudnimg committed:
Limit build number to uint8_t (#49)

Currently the MMU only sends one byte so there is no need
in storing the version with two bytes

gudnimg commented:
First obvious bug I see: (EDIT: its FIXED)

Reproduce steps
1. Start a print and wait for mesh leveling to complete
2. When the loading to nozzle starts, insert NO filament
3. FINDA DIDNT TRIGGER error appears
4. **ISSUE:** Static characters are not on the screen

gudnimg committed:
Fix an issue with static characters not rendering

lcd_pause_print calls return_to_status_screen
and we don't need to do that here.

gudnimg committed:
Fix parking when not in manage_response

gudnimg committed:
Merge branch 'PFW-1375' into PFW-1376

This is needed to test PFW-1376 properly.

With PFW-1375, I get the communication timout error on the LCD

gudnimg commented:
I had to merge merge https://github.com/DRracer/Prusa-Firmware/pull/46 into this PR in order to test it properly.

gudnimg committed:
Fix issue where G1 does not sync its move

gudnimg committed:
Add Keepalive state to MMU error screen

This can be useful to see if something is unintentionally changing the state

gudnimg committed:
Fix issue where error screen is not dismissed upon reconnection

gudnimg commented:
Marking as draft again. Will try to find another solution to this. https://github.com/DRracer/Prusa-Firmware/pull/48/commits/c68a02ee69a8f44be507987748d2232137494a1a 

EDIT: fixed

gudnimg committed:
Improve pausing and parking due to MMU error

Do it in multiple parts to take into account when an error appears nested inside plan_buffer_line

When mmu_loop is called inside plan_buffer_line
while its executing a G1 gcode it is problematic to park the extruder.
We only set the state to PAUSE_PRINT and then on the next iteration of mmu_loop in the top-level loop()
we've exited plan_buffer_line and we can safely park the extruder

Previously this was problematic due the G1 gcode not being done executing

gudnimg committed:
Fix typo

gudnimg committed:
Separate Unparking from Resuming print state

vintagepc commented on Firmware/mmu2.cpp:
```diff
@@ -181,6 +182,53 @@ void MMU2::PowerOn(){
+    {
+    case MMUErrorPrintStates::PAUSE_PRINT:
+        start_pause_print = _millis();
+        stop_and_save_print_to_ram(0.0, -default_retraction);
```
I still don't follow the technical reason for needing/implementing a different and completely orthogonal pause state here. I don't think you ever want an error condition to behave like a normal "paused" print, that's going to open up a whole complicated mess of potential issues relating to the tune menu, as well as users potentially doing a remote-resume from Octoprint without attending to the actual problem. IMO all MMU errors should follow the same pausing pathway in order to keep the code simpler and reduce future bugs/corner cases.

gudnimg replied:
I’m confused how to proceed. I feel like we should at minimum set `isPrintPaused` to `true` etc. So any logic in the `loop()` doesnt do anything unexpected. 

> IMO all MMU errors should follow the same pausing pathway

Do I understand correctly you mean the code shouldn’t care whether we are printing or not?

gudnimg replied:
I took out the Resume and Pause states and did quick test. Didn't see any issues. [4cb9f78](https://github.com/DRracer/Prusa-Firmware/pull/48/commits/4cb9f7873698b872278db5c8861ac8378221e327)

vintagepc replied:
Yes - in effect all errors are "while printing" or doing something to prepare to print such as a load.

gudnimg replied:
Up to now we’ve only been parking while in `manage_response()` which is its own loop and blocks any gcodes from processing during parking. The errors which can happen at any time like the Communication timeout or Protocol error are different in that they happen in the top level `loop()` and we can’t block queued gcodes moves without emptying the planner/command buffer. I’m seeing usually around 6 to 12 commands queued up.

gudnimg committed:
Remove the "Pause" and "Resume" print states.

gudnimg commented:
Parking doesn't work when communication timeout happens. Will look into it. Also stopping a print will cause the printer to crash its axis (its trying to execute some gcode)
![image](https://user-images.githubusercontent.com/8218499/178501926-64c6cb1b-beb7-4343-bf0a-74b0912c5ae3.png)

gudnimg committed:
Revert "Remove the "Pause" and "Resume" print states."

This reverts commit 4cb9f7873698b872278db5c8861ac8378221e327.

gudnimg commented:
I reverted the change since it does not work properly. I tried emptying the planner and command queue just before parking, it didn't help (printer just ignored the parking move completely). I think it requires a full main loop() to work as Yuri mentioned. And we probably need stuff in `stop_and_save_print_to_ram` to save the position in the SD card file (or USB line). Either way this isn't as easy as I thought :)

gudnimg commented:
TODOs

- [x] Add TODO in the code to remind us to block any gcode parsing while parked. We’ll look at it after thermal model branch is merged?
- [x] Reduce `lcd_resume_print`. We don’t need all that stuff, just restore the command queue.

gudnimg committed:
Simplify lcd_resume_print call

And add a TODO comment

gudnimg committed:
Fix an issue where if the next error is the same as the last error, the printer wouldn't park.

gudnimg commented:
I implemented the TODOs and fixed one bug I found while testing.

gudnimg committed:
Add Octoprint Resumed

I'm assuming we need this?

vintagepc committed:
Magic numbers are bad...

DRracer committed:
Merge branch 'mk3-mmu' into PFW-1375

DRracer committed:
Merge pull request #46 from gudnimg/PFW-1375

PFW-1375 MMU error reported by MK3S does not appear on LCD

Commit:
Fix protocol error recovery

Communication timeout and Protocol Errors are now distinguished.
In case of a Protocol Error, the printer waits for heartBeatTimeout to allow filling up the input UART buffer (we expect the MMU still produces some bytes).
Once the timeout elapsed, the input UART buffer is cleared and a new Start Sequence is initiated.

Commit:
Bump MMU FW version to 2.0.19 to match/enforce the necessary compatibility level

DRracer committed:
Merge pull request #51 from DRracer/fix-protocol-error-recovery

Fix protocol error recovery

gudnimg committed:
Merge branch 'mk3-mmu' into PFW-1376

gudnimg commented:
Fixed the merge conflict. The diff is a lot cleaner now.

gudnimg committed:
fixup

gudnimg committed:
Slightly shorter code, no functional difference
